ceph_deploy/gatherkeys.py

```diff
@@ -1,84 +1,254 @@
+        'mds' : 'client.bootstrap-mds',
+        'osd' : 'client.bootstrap-osd',
+        'rgw' : 'client.bootstrap-rgw',
+        'mon' : 'mon.'
+    }
+    return ident_dict.get(keytype, None)
+
+
+def keytype_capabilities(keytype):
+    """
+    Get the capabilities of a keyring from keyring type.
+    """
+    cap_dict = {
+        'admin' : [
+            'osd', 'allow *',
+            'mds', 'allow *',
+            'mon', 'allow *'
+            ],
+        'mds' : [
+            'mon', 'allow profile bootstrap-mds'
+            ],
+        'osd' : [
+            'mon', 'allow profile bootstrap-osd'
+            ],
+        'rgw': [
+            'mon', 'allow profile bootstrap-rgw'
+            ]
+        }
+    return cap_dict.get(keytype, None)
+
 
-            if key is not None:
-                LOG.debug('Got %s key from %s.', topath, hostname)
-                with file(topath, 'w') as f:
-                    f.write(key)
-                    return True
-            distro.conn.exit()
-            LOG.warning('Unable to find %s on %s', filepath, hostname)
-    return False
+def gatherkeys_missing(args, distro, rlogger, keypath, keytype, dest_dir):
+    """
+    Get or create the keyring from the mon using the mon keyring by keytype and
+    copy to dest_dir
+    """
+    arguments = [
+        '/usr/bin/ceph',
+        '--connect-timeout=25',
+        '--cluster={cluster}'.format(
+            cluster=args.cluster),
+        '--name', 'mon.',
+        '--keyring={keypath}'.format(
+            keypath=keypath),
+        'auth', 'get-or-create',
+        ]
+    identity = keytype_identity(keytype)
+    if identity is None:
+        raise RuntimeError('Could not find identity for keytype:%s' % keytype)
+    arguments.append(identity)
+    capabilites = keytype_capabilities(keytype)
+    if capabilites is None:
+        raise RuntimeError('Could not find capabilites for keytype:%s' % keytype)
+    arguments.extend(capabilites)
+    out, err, code = remoto.process.check(
+        distro.conn,
+        arguments
+        )
+    if code != 0:
+        rlogger.error('"ceph auth get-or-create for keytype %s returned %s' % (keytype, code))
+        for line in err:
+            rlogger.debug(line)
+        return False
+    keyring_name_local = keytype_path_to(args, keytype)
+    keyring_path_local = os.path.join(dest_dir, keyring_name_local)
+    with file(keyring_path_local, 'w') as f:
+        for line in out:
+            f.write(line + '\n')
+    return True
+
+
+def gatherkeys_with_mon(args, host, dest_dir):
+    """
+    Connect to mon and gather keys if mon is in quorum.
+    """
+    distro = hosts.get(host, username=args.username)
+    dir_keytype_mon = ceph_deploy.util.paths.mon.path(args.cluster, host)
+    path_keytype_mon = "%s/keyring" % (dir_keytype_mon)
+    mon_key = distro.conn.remote_module.get_file(path_keytype_mon)
+    if mon_key is None:
+        LOG.warning("No mon key found. Is '%s' a mon node" % (host))
```

Comment:
this log statement is a bit confusing. It is asking a question about a host being a mon or is it stating that the mon key is not there? Log statements should be in imperative mode. Something along the lines of:

    LOG.warning("No mon key found in host: %s", host)